### PR TITLE
DataFrame count method

### LINKED ISSUE #1926: Add `len` method to `DataFrame`
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
(This section helps Arrow developers understand the context and *why* for this feature, in addition to  the *what*)
As a user of DataFusion i wouldl like a method to get the length of a `DataFrame`

**Describe the solution you'd like**
a new method `len` on the `DataFrame` trait

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by Jefffrey**
# Which issue does this PR close?

Closes #1926.

# Rationale for this change

# What changes are included in this PR?

New `count()` method on DataFrame

# Are these changes tested?

New unit test

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/core/src/dataframe.rs [hunk 1]
```diff
@@ -361,6 +364,35 @@ impl DataFrame {
...
+    /// # use datafusion::error::Result;
+    /// # #[tokio::main]
+    /// # async fn main() -> Result<()> {
+    /// let ctx = SessionContext::new();
+    /// let df = ctx.read_csv("tests/data/example.csv", CsvReadOptions::new()).await?;
+    /// let length = df.len().await?;
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub async fn len(self) -> Result<usize> {
```
Comment (viirya) [inline datafusion/core/src/dataframe.rs:380]: Hmm, could it be `count`? For DataFrame-like API, there is commonly `count` for this.

Comment (Jefffrey) [reply datafusion/core/src/dataframe.rs]: yeah that makes sense, changed

**Review by andygrove — APPROVED** [review]
Thanks @Jefffrey. Could you also update the user guide?

**Review by viirya — APPROVED** [review]

**Comment by andygrove** [comment]
clippy failures can be fixed by merging latest from master

File: datafusion/core/src/dataframe.rs [hunk 2]
```diff
@@ -361,6 +363,39 @@ impl DataFrame {
...
+    /// let df = ctx.read_csv("tests/data/example.csv", CsvReadOptions::new()).await?;
+    /// let count = df.count().await?;
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub async fn count(self) -> Result<usize> {
+        let rows = self
+            .aggregate(
+                vec![],
+                vec![datafusion_expr::count(Expr::Literal(ScalarValue::Null))],
```
Comment (tustvold) [inline datafusion/core/src/dataframe.rs:384]: This was actually working by accident as a result of a quirk of NullArray whereby it doesn't have a null buffer despite all values being null. The count reported by this query should be 0 as only non-null values are counted.

Fix in https://github.com/apache/arrow-datafusion/pull/5612/files#diff-932cfd7271917561280a69edabb35cfd109d22ae736f77e85adcf63455918121R630

**Pull request merged**